keysign/send.py

```diff
@@ -84,72 +83,155 @@ def __init__(self, builder=None):
+        self.klw.ib.hide()
         self.key = key
         log.info("Activated key %r", key)
         ####
         # Start network services
-        discovery_list = []
-        self.a_offer = AvahiHTTPOffer(self.key)
-        a_data = self.a_offer.start()
-        discovery_list.append(a_data)
-        bt_data = None
-        if BluetoothOffer:
-            self.bt_offer = BluetoothOffer(self.key)
-            bt_data = self.bt_offer.allocate_code()
-            if bt_data:
-                discovery_list.append(bt_data)
+        self.klw.code_spinner.start()
+        if self.internet_option:
+            #self.kpw.internet_spinner.start()
+            # After 10 seconds without a wormhole code we display an info bar
+            timer = 10
+            self.notify = reactor.callLater(timer, self.slow_connection)
+            self.offer = Offer(self.key)
+            try:
+                info = yield self.offer.allocate_code(worm=True)
+            except ServerConnectionError:
+                # We are without a working Internet connection so we stop the previously
+                # activated Avahi server and we display an infobar
+                self._deactivate_timer()
+                self.offer.stop_avahi()
+                self.offer = None
+                self.klw.code_spinner.stop()
+                self.no_connection()
+                return
+
+            code, discovery_data = info
+            self.create_keypresent(code, discovery_data)
+            defers = self.offer.start()
+            for de in defers:
+                # TODO handle errors here?
+                de.addCallback(self._received)
         else:
-            log.debug("No BluetoothOffer: %r", BluetoothOffer)
-        discovery_data = ";".join(discovery_list)
-        if bt_data:
-            self.bt_offer.start().addCallback(self._restart_bluetooth)
+            self.offer = Offer(self.key)
+            info = yield self.offer.allocate_code(worm=False)
+            code, discovery_data = info
+            self.create_keypresent(code, discovery_data)
+            defers = self.offer.start()
+            for de in defers:
+                de.addCallback(self._received)
+
+    def _received(self, start_data):
+        success, message = start_data
+        if message and type(message) == LonelyError:
+            # This only means that we closed wormhole before a transfer
+            pass
+        elif message and message == "Back":
+            # Simply the return of the back button, no errors here
+            pass
+        elif message and type(message) == ServerConnectionError:
+            self._deactivate_timer()
+            self.deactivate()
+            self.klw.code_spinner.stop()
+            self.no_connection()
         else:
-            log.info("Bluetooth as been skipped")
+            self.show_result(success, message)
+
+    def slow_connection(self):
+        self.klw.label_ib.set_label("Very slow Internet connection!")
```

Comment:
Is this a user facing string which we need to translate?

Also maybe we can come up with something more accurate like "still trying to get a connection to the Internet. It appears to be slow or unavailable." or something. But I'm not insisting here.

Reply:
Done with e5bfb2d